compiler/src/iree/compiler/Codegen/Common/TransformDialectInterpreterPass.cpp

```diff
@@ -0,0 +1,175 @@
+
+using namespace mlir;
+
+namespace {
+
+/// Pass declaration.
+/// Interpreter pass that applies transform dialect ops for codegen.
+/// This needs to be its own pass because the registration mechanism and ops
+/// available are different than for other interpreters.
+class TransformDialectInterpreter
+    : public PassWrapper<TransformDialectInterpreter, Pass> {
+ public:
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TransformDialectInterpreter)
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    // TODO: this is only necessary to make registry subset happy when running
+    // the lowering to LLVM. The lowering should be changed to stop using the
+    // nested pass manager and this will go away.
+
+    // clang-format off
+    registry.insert<mlir::iree_compiler::IREE::LinalgExt::IREELinalgExtDialect,
+                    mlir::iree_compiler::IREE::Flow::FlowDialect,
+                    arith::ArithmeticDialect,
+                    AffineDialect,
+                    bufferization::BufferizationDialect,
+                    func::FuncDialect,
+                    gpu::GPUDialect,
+                    linalg::LinalgDialect,
+                    linalg::transform::LinalgTransformDialect,
+                    LLVM::LLVMDialect,
+                    pdl::PDLDialect,
+                    pdl_interp::PDLInterpDialect,
+                    scf::SCFDialect,
+                    tensor::TensorDialect,
+                    vector::VectorDialect
+        // clang-format on
+        >();
+
+    // TODO: these should be registered by the extension instead, but there is
+    // no support for it in core currently.
+    arith::registerBufferizableOpInterfaceExternalModels(registry);
+    linalg::registerBufferizableOpInterfaceExternalModels(registry);
+    scf::registerBufferizableOpInterfaceExternalModels(registry);
+    bufferization::func_ext::registerBufferizableOpInterfaceExternalModels(
+        registry);
+    tensor::registerBufferizableOpInterfaceExternalModels(registry);
+    vector::registerBufferizableOpInterfaceExternalModels(registry);
+  }
+
+  StringRef getArgument() const override {
+    return "iree-transform-dialect-interpreter";
+  }
+
+  StringRef getDescription() const override {
+    return "apply transform dialect operations one by one";
+  }
+
+  bool canScheduleOn(RegisteredOperationName name) const override {
+    return true;
+  }
+
+  TransformDialectInterpreter(StringRef transformFileName = StringRef()) {
+    this->transformFileName = transformFileName.str();
+  }
+  TransformDialectInterpreter(const TransformDialectInterpreter &pass) {
+    this->transformFileName = pass.transformFileName;
+    // TODO: if we really don't like shared_ptr, we could also clone the
+    // transformModule here.
+    sharedTransformModule = pass.sharedTransformModule;
+  }
+
+  LogicalResult initialize(MLIRContext *context) override {
+    OwningOpRef<ModuleOp> module;
```

Comment:
Why `OwningOpRef`?

Reply:
So that when the shared_ptr releases its last use, the module is destructed.